### ISSUE TITLE
Problems with optgroups and filter

### ISSUE THREAD

**Issue by Generaler** [davidstutz/bootstrap-multiselect#680]
First of all, thanks for great job with Bootstrap-multiselect. It's really awsome!
I had a couple of problems and maybe my solutions will help someone.

1. When you use "enableClickableOptGroups" with "enableCollapsibleOptGroups", there is little problem with check/uncheck options. If you check all optgroups and then uncheck any option in any optgroup except first one, plugin unchecks not only target optrgoup checkbox (with target option), it unchecks all prevous optgroups checkboxes even if all their options are checked. You can see it on exaple page.
To fix it you need just to make a little modification in bootstrap-multiselect.js.
<br><br>change this:
<br><code>
 n.prevAll('.multiselect-group').find('input').prop('checked', i);
</code>
<br><br>to this:
<br><code>
 n.prevAll('.multiselect-group:first').find('input').prop('checked', i);
</code>
<br><br>
2. Second issue concern filter. If you use it with "enableCollapsibleOptGroups", when you  type something  you need to expand optgroups to see what you found. To fix it you need to add one string to bootstrap-multiselect.js in buildFilter function to toggle class block (// Toggle current element (group or group item) according to showElement boolean.):
<br><br><code>
$(element).toggle(showElement).toggleClass('filter-visible', showElement);
</code>
<br><br>
And then add to your CSS new class .filter-visible with this:
<br><br>
<code>
display: list-item !important;
</code><br><br>
Now when you type something in the filter input, you will see results without expanding optgroups.
Sorry for my bad English. Just wanted to share my experience of using this amazing plugin.

**Comment by HarishUkirade** [comment]
There is another issue related with first issue:
e.g. I have same item in in 2 - 3 groups, if I uncheck item from one group, same item which is available from other group should be unchecked. 

in this case below line wont work:
n.prevAll('.multiselect-group:first').find('input').prop('checked', i);

**Comment by mabgfounder** [comment]
Generaler solution to option one did not completely solve the problem for me. If there were more than two groups then I still have multiple groups getting checked/unchecked. 

However, I think Generaler correctly identified the problem area of code. However, I think the issue is the author intended for that code to only run when a non-group checkbox changed. However, it runs when any checkbox changes. This causes the prevUntil/nextUntil/prevAll to cast a wider net than the author intended. Earlier in the code there is another event handler for specifically when the 'li.multiselect-group > a > input' checkboxes change. 

I had good results when I changed this:
```
$("li` input", this.$ul).on("change", $.proxy(function(t) {
```
to this:
```
$("li:not(.multiselect-group) > a > input", this.$ul).on("change", $.proxy(function (t) {
```

I have not checked it with HarishUkirade where the same option occurs in multiple groups.

**Comment by mabgfounder** [comment]
I also had the problem described in point 1. I don't especially like my fix but it seems to work so I have attached a patch. This same problem is also reported in #655.

[bootstrap-multiselect.js.655n680.patch.txt](https://github.com/davidstutz/bootstrap-multiselect/files/108326/bootstrap-multiselect.js.655n680.patch.txt)

**Comment by davidstutz** [comment]
See #655, will have a look into it.

**Comment by davidstutz** [comment]
Point 1 should be fixed as confirmed in #655. Problem 2 should also be fixed.
